server/channels/app/authentication.go

```diff
@@ -59,7 +59,24 @@ func (a *App) IsPasswordValid(rctx request.CTX, password string) *model.AppError
 	return nil
 }
 
-func (a *App) CheckPasswordAndAllCriteria(rctx request.CTX, user *model.User, password string, mfaToken string) *model.AppError {
+func (a *App) CheckPasswordAndAllCriteria(rctx request.CTX, u *model.User, password string, mfaToken string) *model.AppError {
+	// MM-37585
+	// The user param will have stale info. in case of concurrent failed login attempts by the same user.
+	// So, lock and fetch the updated attempts count.
+	a.ch.loginAttemptsMut.Lock()
+	defer a.ch.loginAttemptsMut.Unlock()
+
+	a.Srv().Store().User().InvalidateProfileCacheForUser(u.Id)
+	user, err := a.GetUser(u.Id)
+	if err != nil {
+		if err.Id != MissingAccountError {
+			err.StatusCode = http.StatusInternalServerError
+			return err
+		}
+		err.StatusCode = http.StatusBadRequest
+		return err
+	}
```

Comment:
What is the purpose of this?

Reply:
The purpose of this and the above cache invalidation is to get an updated count of failure attempts. Since, the user is passed as a param, multiple goroutines may already have fetched this user before ever calling this function and thus will have stale data about the loginAttempts.

Reply:
Thanks. In that case, could we change the function signature to just take the userID instead? That should force the consumer to fetch the user details and prevent any accidental stale usage of user attributes.

Reply:
> That should force the consumer to fetch the user details

You mean _not_ force the consumer to fetch the user details? As consumer of this function are already fetching the user details and passing it to this function.

The signature change would only help in making it a bit clearer that this function will fetch the user details on its own. However, if I'm understanding correctly, this doesn't change anything on the consumer side, right?

Reply:
Ah yes, I meant, it should force the _implementor_ to fetch the user details.